sdk/src/main/java/io/opentelemetry/sdk/trace/RecordEventsReadableSpan.java

```diff
@@ -362,24 +363,23 @@ public void addEvent(io.opentelemetry.trace.Event event, long timestamp) {
     addTimedEvent(TimedEvent.create(timestamp, event));
   }
 
-  static Map<String, AttributeValue> copyAndLimitAttributes(
-      Map<String, AttributeValue> attributes, int limit) {
-    if (attributes.isEmpty()) {
-      return Collections.emptyMap();
-    }
-
-    if (attributes.size() <= limit) {
-      return Collections.unmodifiableMap(new HashMap<>(attributes));
-    }
-
-    Map<String, AttributeValue> temp = new HashMap<>();
-    for (Map.Entry<String, AttributeValue> entry : attributes.entrySet()) {
-      if (temp.size() < limit) {
-        temp.put(entry.getKey(), entry.getValue());
-      }
+  static Attributes copyAndLimitAttributes(final Attributes attributes, final int limit) {
+    if (attributes.isEmpty() || attributes.size() <= limit) {
+      return attributes;
     }
 
-    return Collections.unmodifiableMap(temp);
+    final Attributes.Builder temp = Attributes.newBuilder();
+    final AtomicInteger added = new AtomicInteger();
```

Comment:
why atomic? I think I know why because you need to change it in the lambda. You can have a size() on the builder maybe? Maybe create an issue because this smells a bit :)

Reply:
yes, entirely because it needed to be updated in the inner class. `size()` on the builder isn't a bad idea at all, though. 

I will note that we're going to need a more sophisticated builder-like class to handle the attributes on the active span itself, so maybe we can roll the cleanup into that change?